### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Stokalas**
**Authentication using cookies**
- Login
- Register
- Certain pages only reachable when logged in
- When rerouted to Login after loggin in continues to the same page
-  Different Jumbotron in Index according to whether user is logged in.

### CODE REVIEW

File: src/Pages/Register.cshtml.cs [hunk 1]
```diff
@@ -1,16 +1,62 @@
...
+
+        public async Task<IActionResult> OnPost()
         {
+            if (!Password.Equals(PasswordConfirm))
+            {
+                ErrorMessage = "Passwords did not match!";
+                return Page();
+            }
+
+            var Id = UserAuth.Registration(Email, Password);
```
Comment (ciliakas) [inline src/Pages/Register.cshtml.cs]: 'Id' should probably start with a lowercase letter

File: src/Pages/Register.cshtml.cs [hunk 2]
```diff
@@ -1,16 +1,62 @@
...
+                 {
+                     new Claim(ClaimTypes.UserData, Id.ToString()),
+                     new Claim(ClaimTypes.Email, Email)
+                 };
+                 var claimsIdentity = new ClaimsIdentity(claims, "Login");
+                 await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme,
+                     new ClaimsPrincipal(claimsIdentity));
+ 
+                 return Redirect("/Index");
+            };
```
Comment (ciliakas) [inline src/Pages/Register.cshtml.cs]: Could you explain how the registration works?  The things that caught my eye were 'async Task' and 'Claim',

File: src/Pages/logIn.cshtml.cs [hunk 3]
```diff
@@ -1,31 +1,49 @@
...
-        public void OnPost()
+
+        public void OnGet()
         {
-            if (UserAuth.Login(Email, Password))
+            ErrorMessage = "";
+        }
+        public async Task<IActionResult> OnPost(string ReturnUrl)
+        {
+            var Id = UserAuth.Login(Email, Password);
```
Comment (ciliakas) [inline src/Pages/logIn.cshtml.cs]: 'Id' and 'ReturnUrl' should probably start with lowercase

**Review by ciliakas — CHANGES_REQUESTED** [review]
Everything works, the only small glitch that I caught was that if you have your login information saved in browser, if you click register the same info is already filled in into the registration form.
Other small issues in the comments

**Comment by Stokalas** [comment]
**Added fixes:**
- Renamed id and returnUrl variables/parameters
- Turned off autocomplete on Register form
- When trying to access Login/Register when logged in (by link for example: `https://localhost:44330/Login` reroutes to `/Index`

File: src/Pages/Register.cshtml.cs (hunk 2 above)
Comment (Stokalas) [reply src/Pages/Register.cshtml.cs]: - Citing: 
> An async method runs synchronously until it reaches its first await expression, at which point the method is suspended until the awaited task is complete. "

- Claims are data that is saved in cookie and can be accessed as long as person logged in.

- Creating list of claims (data saved in cookie):
` var claims = new List<Claim>
                 {
                     new Claim(ClaimTypes.UserData, Id.ToString()),
                     new Claim(ClaimTypes.Email, Email)
                 };`

- Variable with specified claims, authentication type.
`var claimsIdentity = new ClaimsIdentity(claims, "Login");`

- Creation of cookie (boilerplate code not sure about each step of it):
` await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme,
                     new ClaimsPrincipal(claimsIdentity));`


- And finally: 

> "IActionResult type. The IActionResult return type is appropriate when multiple ActionResult return types are possible in an action. The ActionResult types represent various HTTP status codes. Any non-abstract class deriving from ActionResult qualifies as a valid return type."

**_Quite a bit to read, but I tried to be as informative as I could._** 🙂

Comment (Stokalas) [reply src/Pages/Register.cshtml.cs]: I skipped some parts since I think they are understandable.

Comment (ciliakas) [reply src/Pages/Register.cshtml.cs]: Thanks for the in depth answer, clarified this method alot!

**Review by ciliakas — APPROVED** [review]
Approving after fixes

**Pull request merged**